Delhi Daily Climate — Pandas & NumPy Data Analysis Assignment
AI use case: Climate, Energy & Sustainability — heatwave forecasting & early warning. Dataset: DailyDelhiClimateTest.csv (114 daily records: date, meantemp, humidity, wind_speed, meanpressure).

READ THE DATASET AND IMPORT LIBRARIES

In [13]:
import pandas as pd
import numpy as np
df = pd.read_excel(r"C:\Users\kjsce_mech45\Desktop\RAI_EED_Manav\ICRISAT Weather 1978 to 2018.xlsx")


FIND THE MEAN OF MAXIMUM TEMPERATURE RECORDED

In [14]:
data_Max = df["MaxT"].dropna().tolist()

total = 0

for x in data_Max:
    total = total + x

n = len(data_Max)
mean = total / n

print("Mean of Maximum Temperature =", mean)

Mean of Maximum Temperature = 32.05580690769607


The average maximum temperature recorded in the dataset is approximately 32.06 degree C, indicating that the study area generally experiences a warm climate.

FIND THE MEDIAN OF RELATIVE HUMIDITY IN THE MORNING

In [15]:
data_Humid = df["RH1"].dropna().tolist()
n = len(data_Humid)

# Bubble Sort
for i in range(n):
    for j in range(0, n - i - 1):
        if data_Humid[j] > data_Humid[j + 1]:
            temp = data_Humid[j]
            data_Humid[j] = data_Humid[j + 1]
            data_Humid[j + 1] = temp


if n % 2 == 0:
    median = (data_Humid[n // 2 - 1] + data_Humid[n // 2]) / 2
else:
    median = data_Humid[n // 2]
print("Median of Relative Humidity in Morning =", median)


Median of Relative Humidity in Morning = 87


The median morning relative humidity is 87%, which means that half of the recorded RH1 values are below 87% and the other half are above it. This indicates that the mornings are generally associated with high relative humidity.

FIND THE MODE OF RELATIVE HUMIDITY IN AFTERNOON

In [16]:
data_HumidA = df["RH2"].dropna().tolist()

frequency = {}
for x in data_HumidA:
    if x in frequency:
        frequency[x] = frequency[x] + 1
    else:
        frequency[x] = 1

max_count = 0
mode = None

for x in frequency:

    if frequency[x] > max_count:

        max_count = frequency[x]
        mode = x


print("Mode of Relative Humidity in Afternoon =", mode)


Mode of Relative Humidity in Afternoon = 29.0


The most frequently occurring afternoon relative humidity is 29%. This suggests that relatively low humidity conditions occur frequently during the afternoon.




VARIANCE OF THE RADIATION

In [17]:
data_Rad = df["Radiation"].dropna().tolist()

n = len(data_Rad)
total = 0

for x in data_Rad:
    total = total + x

mean_rad = total / n


# Calculate variance
sum_squared_difference = 0

for x in data_Rad:
    difference = x - mean_rad
    sum_squared_difference = (sum_squared_difference + difference * difference)

variance = sum_squared_difference /n
print("Variance of Radiation =", variance)

Variance of Radiation = 20.32170357510379


The variance of radiation is 20.32, indicating that the radiation values are spread around their mean rather than remaining constant. This variation can occur due to changes in weather conditions, cloud cover, seasons, and daily atmospheric conditions.

STANDARD DEVIATION

In [19]:
Std= np.sqrt(variance)
print(mean_rad)
print("Standard Deviation of Radiation =", Std)

17.87702666307564
Standard Deviation of Radiation = 4.507960023680755


The standard deviation of approximately 4.51 indicates the typical deviation of radiation values from the mean radiation value of 17.88. Therefore, radiation measurements show a moderate level of variability and are not concentrated at exactly one value.

INTERQUARTILE RANGE AND OUTLIERS

In [20]:
attribute = "MaxT"
data = df[attribute].dropna()
# Calculate Quartiles
Q1 = data.quantile(0.25)
Q2 = data.quantile(0.50)
Q3 = data.quantile(0.75)

# Calculate IQR
IQR = Q3 - Q1
print("Q1 =", Q1)
print("Q2 =", Q2)
print("Q3 =", Q3)
print("IQR =", IQR)

lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

print("\nLower Bound =", lower_bound)
print("Upper Bound =", upper_bound)

outliers = data[(data < lower_bound) | (data > upper_bound)]

print("Number of Outliers =", outliers.count())

print("Outlier Values:")
print(outliers.tolist())

print("Minimum Extreme Value =", data.min())
print("Maximum Extreme Value =", data.max())


Q1 = 29.0
Q2 = 31.0
Q3 = 35.0
IQR = 6.0

Lower Bound = 20.0
Upper Bound = 44.0
Number of Outliers = 4
Outlier Values:
[17.8, 16.5, 19.0, 19.5]
Minimum Extreme Value = 16.5
Maximum Extreme Value = 43.5


Most of the maximum temperature values lie between 29 degrees and 35 degrees, with an IQR of 6 degrees. There are 4 low-temperature outliers: 17.8 degrees, 16.5 degrees, 19 degrees, and 19.5 degrees, as they fall below the lower bound of 20 degrees. There are no high-temperature outliers, since the maximum value of 43.5 degrees is within the upper bound of 44 degrees.

GROUPED DATA

sturges rule is not used in this case since it is used only upto 200 data observations

In [21]:
number_of_classes = 10

minimum = data.min()
maximum = data.max()

class_width = (maximum - minimum) / number_of_classes

bins = np.linspace(
    minimum,
    maximum,
    number_of_classes + 1
)

grouped_data = pd.cut(
    data,
    bins=bins,
    include_lowest=True
)

frequency = grouped_data.value_counts().sort_index()

frequency_table = pd.DataFrame({"Class Interval": frequency.index.astype(str), "Frequency": frequency.values})


print("GROUPED FREQUENCY DISTRIBUTION")
print(frequency_table)
print("STATISTICAL ANALYSIS")


print("Mean =", data.mean())
print("Mode =", data.mode().tolist())
print("Median =", data.median())
print("Variance =", data.var())
print("Standard Deviation =", data.std())
print("Q1 =", Q1)
print("Q3 =", Q3)
print("Interquartile Range =", IQR)

GROUPED FREQUENCY DISTRIBUTION
   Class Interval  Frequency
0  (16.499, 19.2]          3
1    (19.2, 21.9]         16
2    (21.9, 24.6]        141
3    (24.6, 27.3]       1142
4    (27.3, 30.0]       4463
5    (30.0, 32.7]       3783
6    (32.7, 35.4]       1898
7    (35.4, 38.1]       1718
8    (38.1, 40.8]       1334
9    (40.8, 43.5]        355
STATISTICAL ANALYSIS
Mean = 32.05580690769542
Mode = [30.0]
Median = 31.0
Variance = 16.934586025734667
Standard Deviation = 4.115165370399429
Q1 = 29.0
Q3 = 35.0
Interquartile Range = 6.0
